File: mobile/android/app/src/main/java/com/hoggcountry/trailassistant/scout/ScoutGemmaEngine.java
package com.hoggcountry.trailassistant.scout;

/**
 * Native boundary for the on-device Gemma engine.
 *
 * This interface is the ONLY place that should know about a model runtime
 * (LiteRT-LM / MediaPipe LLM Inference) or Gemma weights. ScoutGemmaPlugin
 * talks to this interface and never to a runtime directly, so the JS contract in
 * mobile/src/lib/scout/capacitor-gemma-bridge.ts stays stable while the engine
 * is swapped from the honest "unavailable" stub to a real implementation.
 *
 * Integration: ScoutGemmaPlugin.createEngine() tries a local implementation
 * first and falls back to the unavailable stub when the runtime or verified
 * model is missing.
 */
public interface ScoutGemmaEngine {

    /**
     * True when this engine can attempt generation (runtime linked + verified
     * weights present). Runtime initialization readiness is reported separately by
     * {@link #isRuntimeReady()} because loading LiteRT-LM is intentionally kept off
     * the Capacitor/UI thread.
     */
    boolean isAvailable();

    /** True only after the native LiteRT-LM engine has initialized successfully. */
    default boolean isRuntimeReady() {
        return false;
    }

    /**
     * Returns the model this build targets. Implementations may return the
     * target descriptor even when weights are absent (informational), or null
     * when nothing is configured.
     */
    ScoutGemmaModelInfo describeModel();

    /**
     * Runs a single on-device completion, streaming incremental text to {@code sink}
     * as it is produced and returning the full result when generation completes.
     *
     * @param sink receives each incremental text chunk during generation; may be
     *             {@code null} for a non-streaming call (the full text is still
     *             returned). Implementations must tolerate a null sink.
     * @throws ScoutGemmaUnavailableException when the engine cannot generate.
     *         Implementations must never return fabricated text.
     */
    GenerateResult generate(String prompt, String systemContext, int maxTokens, TokenSink sink)
            throws ScoutGemmaUnavailableException;

    /**
     * Eagerly initializes the underlying runtime so the FIRST {@link #generate}
     * call doesn't pay (or risk) the heavy, sometimes-flaky lazy init. Returns only
     * after initialization has succeeded; failures are surfaced honestly to the
     * caller instead of reporting a false-positive warm-up.
     */
    default boolean warmUp() throws ScoutGemmaUnavailableException {
        if (!isAvailable()) {
            throw new ScoutGemmaUnavailableException("On-device Gemma is not available.");
        }
        return isRuntimeReady();
    }

    /**
     * Cooperatively cancels the active generation, if any. This must be callable
     * from outside the serial inference executor; queueing it behind generation
     * would make cancellation ineffective.
     */
    default boolean cancelGeneration() {
        return false;
    }

    /** Receives incremental text chunks as the engine streams a response. */
    interface TokenSink {
        void onToken(String chunk);
    }

    /**
     * Releases any native resources held by this engine instance.
     * Safe to call multiple times; subsequent calls are no-ops.
     * Must be called when swapping engines or when the plugin is destroyed.
     */
    void close();

    /** Mirrors the JS generate() return shape: { text, truncated }. */
    final class GenerateResult {
        public final String text;
        public final boolean truncated;

        public GenerateResult(String text, boolean truncated) {
            this.text = text;
            this.truncated = truncated;
        }
    }
}

File: mobile/ios/App/App/scout/ScoutGemmaEngine.swift
import Foundation

/// Native boundary for the on-device Gemma engine on iOS — the Swift mirror of
/// the Android `ScoutGemmaEngine` interface. This is the ONLY place that should
/// know about a model runtime (LiteRT-LM). `ScoutGemmaPlugin` talks to this
/// protocol and never to the runtime directly, so the JS contract in
/// `mobile/src/lib/scout/capacitor-gemma-bridge.ts` stays identical across
/// Android and iOS.
protocol ScoutGemmaEngine {
    /// True only when a runtime is loaded and model weights are usable.
    var isAvailable: Bool { get }

    /// The model this build targets. May return the target descriptor even when
    /// weights are absent (informational), or nil when nothing is configured.
    func describeModel() -> ScoutGemmaModelInfo?

    /// Eagerly initializes the native runtime and verifies that the model can be
    /// loaded. This is intentionally awaited: callers must not report the
    /// runtime as warmed until LiteRT-LM initialization has actually completed.
    func warmUp() async throws

    /// Runs a single on-device completion. Throws `ScoutGemmaError.unavailable`
    /// when the engine cannot generate — implementations must never return
    /// fabricated text.
    func generate(prompt: String, systemContext: String, maxTokens: Int) async throws -> GenerateResult
}

/// Mirrors the JS `generate()` return shape: `{ text, truncated }`.
struct GenerateResult {
    let text: String
    let truncated: Bool
}

/// Descriptor `{ tier, modelId, maxContextTokens }`, mirrors JS `GemmaModelDescriptor`.
struct ScoutGemmaModelInfo {
    let tier: String
    let modelId: String
    let maxContextTokens: Int
}

enum ScoutGemmaError: LocalizedError {
    case unavailable(String)
    case timedOut(seconds: Int)
    case cancelled

    var errorDescription: String? {
        switch self {
        case .unavailable(let message): return message
        case .timedOut(let seconds):
            return "On-device Gemma timed out after \(seconds)s. Close and reopen Scout before trying local AI again."
        case .cancelled:
            return "On-device Gemma generation was cancelled."
        }
    }

    var code: String {
        switch self {
        case .unavailable: return "SCOUT_GEMMA_UNAVAILABLE"
        case .timedOut: return "SCOUT_GEMMA_TIMEOUT"
        case .cancelled: return "SCOUT_GEMMA_CANCELLED"
        }
    }
}

/// Default engine used until the LiteRT-LM Swift package is added and a verified
/// model is on the device. Honestly reports the model as unavailable and never
/// fabricates output, so Gemma-only builds block chat instead of using a
/// paid/cloud model. `describeModel()` still returns the target descriptor so
/// UI/telemetry can show what model this build is built around.
struct UnavailableScoutGemmaEngine: ScoutGemmaEngine {
	static let targetModel = ScoutGemmaModelInfo(
		tier: "balanced",
		modelId: "gemma-4-E2B-it-litert-lm",
		maxContextTokens: 32_768
	)

    var isAvailable: Bool { false }

    func describeModel() -> ScoutGemmaModelInfo? { Self.targetModel }

    func warmUp() async throws {
        throw ScoutGemmaError.unavailable(
            "On-device Gemma cannot warm up until the LiteRT-LM runtime and a verified model are available."
        )
    }

    func generate(prompt: String, systemContext: String, maxTokens: Int) async throws -> GenerateResult {
        throw ScoutGemmaError.unavailable(
            "On-device Gemma engine is not available in this build. Add the LiteRT-LM Swift package and download a verified Gemma 4 model file so the local runtime can load it."
        )
    }
}
